platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.gradle.api.initialization.Settings
+import org.gradle.api.internal.SettingsInternal
+import org.gradle.internal.InternalBuildAdapter
+import org.gradle.internal.concurrent.Stoppable
+import org.gradle.internal.event.ListenerManager
+import org.gradle.internal.service.scopes.Scope
+import org.gradle.internal.service.scopes.ServiceScope
+
+/**
+ * A utility class that provides the name of the build.
+ */
+
+@ServiceScope(Scope.BuildTree::class)
+class BuildNameProvider(val listenerManager: ListenerManager) : InternalBuildAdapter(), Stoppable {
+
+    init {
+        listenerManager.addListener(this)
+    }
+
+    override fun stop() {
+        listenerManager.removeListener(this)
+    }
+
+    var buildName: String? = null
+        get() = field
```

Comment:
❌  This should not be publicly mutable.

We also should throw an exception in case someone tries to get the name before it is available.

It seems that the root project name is always available. So the type should not be nullable.